consensus/reactor.go

```diff
@@ -329,6 +329,19 @@ func (conR *Reactor) Receive(e p2p.Envelope) {
 		}
 		switch msg := msg.(type) {
 		case *ProposalMessage:
+			conR.conS.mtx.RLock()
+			maxBytes := conR.conS.state.ConsensusParams.Block.MaxBytes
+			conR.conS.mtx.RUnlock()
+			if maxBytes == -1 {
+				maxBytes = int64(types.MaxBlockSizeBytes)
+			}
+			totalParts := int64(msg.Proposal.BlockID.PartSetHeader.Total)
+			chunks := (maxBytes-1)/int64(types.BlockPartSizeBytes) + 1
+			if totalParts > chunks {
+				conR.Logger.Error("Rejecting oversized proposal", "peer", e.Src, "height", msg.Proposal.Height)
+				conR.Switch.StopPeerForError(e.Src, ErrProposalTooManyParts)
+				return
+			}
```

Comment:
Nit: we can avoid the explicit return if we invert the clause, and log if it doesn't match:

```suggestion
			if totalParts <= chunks { 
				ps.SetHasProposal(msg.Proposal)
			    conR.conS.peerMsgQueue <- msgInfo{msg, e.Src.ID()}
			} else {
			    conR.Logger.Error("Rejecting oversized proposal", "peer", e.Src, "height", msg.Proposal.Height)
				conR.Switch.StopPeerForError(e.Src, ErrProposalTooManyParts)
			}
```

Reply:
yes i am aligned with keeping this as is to make it clear that this conditional return is an "exceptional" case out of the happy path